homeassistant/__main__.py

```diff
@@ -259,39 +251,33 @@ def open_browser(event):
     hass.start()
     exit_code = int(hass.block_till_stopped())
 
-    if not top_process:
-        sys.exit(exit_code)
     return exit_code
 
 
-def run_hass_process(hass_proc):
-    """Run a child hass process. Returns True if it should be restarted."""
-    requested_stop = threading.Event()
-    hass_proc.daemon = True
-
-    def request_stop(*args):
-        """Request hass stop, *args is for signal handler callback."""
-        requested_stop.set()
-        hass_proc.terminate()
+def try_to_restart():
+    """Attempt to clean up state and start a new homeassistant instance."""
+    # Things should be mostly shut down already at this point, now just try
+    # to clean up things that may have been left behind.
+    sys.stderr.write('Home Assistant attempting to restart.\n')
 
-    try:
-        signal.signal(signal.SIGTERM, request_stop)
-    except ValueError:
-        print('Could not bind to SIGTERM. Are you running in a thread?')
+    # Count remaining threads, ideally there should only be one non-daemonized
+    # thread left (which is us). Nothing we really do with it, but it might be
+    # useful when debugging shutdown/restart issues.
+    nthreads = sum(thread.isAlive() and not thread.isDaemon()
+                   for thread in threading.enumerate())
+    if nthreads > 1:
+        sys.stderr.write("Found {} non-daemonic threads.\n".format(nthreads))
 
-    hass_proc.start()
-    try:
-        hass_proc.join()
-    except KeyboardInterrupt:
-        request_stop()
-        try:
-            hass_proc.join()
-        except KeyboardInterrupt:
-            return False
+    # Try to not leave behind open filedescriptors with the emphasis on try.
+    os.closerange(3, 1024)
```

Comment:
What? cleaning up file descriptors kills the process? I would not have expected that. I'll look into that a bit more to see why that is happening and maybe make a special case for macos. Generally closing fd is a good thing because it also gets rid of open sockets and such.

I can probably revert 705255d then and at least signal (child) processes that are still hanging out in our process group.